pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -326,6 +328,25 @@ type attachDetachController struct {
 
 	// pvcQueue is used to queue pvc objects
 	pvcQueue workqueue.RateLimitingInterface
+
+	// operationTimestamps caches start timestamp of attach/detach for metric recording.
+	// Detailed lifecyle/key for each operation
+	// 1. attach
+	//     key:        string of VolumeToAttach.NodeName + "/" + VolumeToAttach.VolumeName
+	//     start time: the first time "AttachVolume" is called in reconciler in "attachDesiredVolumes"
+	//                 for the VolumeToAttach object
+	//     end time:   after the "VolumeToAttach" has been successfully attached to the desired node
+	//     abort:      The "VolumeToAttach" has been marked "attached", i.e., actualStateOfWorld
```

Comment:
We should also abort when the volume is removed from desired state.  This can happen if the pod is deleted before we finished attaching.

Reply:
good point. another possible scenario is when a pod has been updated to remove a volume. Will enhance to handle those scenarios.